File: app/model/races/Stage.java
package model.races;

import com.fasterxml.jackson.annotation.JsonIgnore;
import model.races.classification.TeamTimeTrialClassification;
import model.races.classification.general.General;
import model.races.classification.general.Mountain;
import model.races.classification.general.Points;
import model.races.classification.mountain.*;
import model.races.classification.sprint.IntermediateSprint;
import model.races.classification.sprint.StageSprint;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Created by Jon on 30/01/2015.
 */
@Entity(name = "stages")
public class Stage implements Comparable<Stage>{

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private long id;
    @ManyToOne
    @JoinColumn(name="raceId")
    @JsonIgnore
    private Race race;
    private String description;
    private int mHC,m1,m2,m3,m4,sInt,stageNumber;
    private boolean isTTT,hasMC,hasRC;

    @Transient
    private General generalClassification;
    @Transient
    private StageSprint stageClassification;
    @Transient
    private TeamTimeTrialClassification tttClassification;
    @Transient
    private Mountain mountainClassification;
    @Transient
    private Points pointsClassification;
    @ElementCollection
    private List<MountainHC> mhcClassificationList;
    @ElementCollection
    private List<MountainFirst> m1ClassificationList;
    @ElementCollection
    private List<MountainSecond> m2ClassificationList;
    @ElementCollection
    private List<MountainThird> m3ClassificationList;
    @ElementCollection
    private List<MountainFourth> m4ClassificationList;
    @ElementCollection
    private List<IntermediateSprint> sIntClassificationList;

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public Race getRace() {
        return race;
    }

    public void setRace(Race race) {
        this.race = race;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getmHC() {
        return mHC;
    }

    public void setmHC(int mHC) {
        this.mHC = mHC;
        if (this.mhcClassificationList == null) {
            this.mhcClassificationList = new ArrayList<MountainHC>(mHC);
        }
    }

    public int getM1() {
        return m1;
    }

    public void setM1(int m1) {
        this.m1 = m1;
        if (this.m1ClassificationList == null) {
            this.m1ClassificationList = new ArrayList<MountainFirst>(m1);
        }
    }

    public int getM2() {
        return m2;
    }

    public void setM2(int m2) {
        this.m2 = m2;
        if (this.m2ClassificationList == null) {
            this.m2ClassificationList = new ArrayList<MountainSecond>(m2);
        }
    }

    public int getM3() {
        return m3;
    }

    public void setM3(int m3) {
        this.m3 = m3;
        if (this.m3ClassificationList == null) {
            this.m3ClassificationList = new ArrayList<MountainThird>(m3);
        }
    }

    public int getM4() {
        return m4;
    }

    public void setM4(int m4) {
        this.m4 = m4;
        if (this.m4ClassificationList == null) {
            this.m4ClassificationList = new ArrayList<MountainFourth>(m4);
        }
    }

    public int getsInt() {
        return sInt;
    }

    public void setsInt(int sInt) {
        this.sInt = sInt;
        if (this.sIntClassificationList == null) {
            this.sIntClassificationList = new ArrayList<IntermediateSprint>(sInt);
        }
    }

    public boolean isTTT() {
        return isTTT;
    }

    public void setTTT(boolean isTTT) {
        this.isTTT = isTTT;
    }

    public int getStageNumber() {
        return stageNumber;
    }

    public void setStageNumber(int stageNumber) {
        this.stageNumber = stageNumber;
    }

    public boolean hasMC() {
        return hasMC;
    }

    public void setMC(boolean hasMC) {
        this.hasMC = hasMC;
    }

    public boolean hasRC() {
        return hasRC;
    }

    public void setRC(boolean hasRC) {
        this.hasRC = hasRC;
    }

    public General getGeneralClassification() {
        return generalClassification;
    }

    public void setGeneralClassification(General generalClassification) {
        this.generalClassification = generalClassification;
    }

    public TeamTimeTrialClassification getTttClassification() {
        return tttClassification;
    }

    public void setTttClassification(TeamTimeTrialClassification tttClassification) {
        this.tttClassification = tttClassification;
    }

    public Mountain getMountainClassification() {
        return mountainClassification;
    }

    public void setMountainClassification(Mountain mountainClassification) {
        this.mountainClassification = mountainClassification;
    }

    public Points getPointsClassification() {
        return pointsClassification;
    }

    public void setPointsClassification(Points pointsClassification) {
        this.pointsClassification = pointsClassification;
    }

    public List<MountainHC> getMhcClassificationList() {
        return mhcClassificationList;
    }

    public void setMhcClassificationList(List<MountainHC> mhcClassificationList) {
        this.mhcClassificationList = mhcClassificationList;
    }

    public List<MountainFirst> getM1ClassificationList() {
        return m1ClassificationList;
    }

    public void setM1ClassificationList(List<MountainFirst> m1ClassificationList) {
        this.m1ClassificationList = m1ClassificationList;
    }

    public List<MountainSecond> getM2ClassificationList() {
        return m2ClassificationList;
    }

    public void setM2ClassificationList(List<MountainSecond> m2ClassificationList) {
        this.m2ClassificationList = m2ClassificationList;
    }

    public List<MountainThird> getM3ClassificationList() {
        return m3ClassificationList;
    }

    public void setM3ClassificationList(List<MountainThird> m3ClassificationList) {
        this.m3ClassificationList = m3ClassificationList;
    }

    public List<MountainFourth> getM4ClassificationList() {
        return m4ClassificationList;
    }

    public void setM4ClassificationList(List<MountainFourth> m4ClassificationList) {
        this.m4ClassificationList = m4ClassificationList;
    }

    public List<IntermediateSprint> getsIntClassificationList() {
        return sIntClassificationList;
    }

    public void setsIntClassificationList(List<IntermediateSprint> sIntClassificationList) {
        this.sIntClassificationList = sIntClassificationList;
    }

    public StageSprint getStageClassification() {
        return stageClassification;
    }

    public void setStageClassification(StageSprint stageClassification) {
        this.stageClassification = stageClassification;
    }

    @Override
    public int compareTo(Stage other) {
        return stageNumber-other.getStageNumber();
    }
}

File: public/javascripts/stage.js
function Stage() {
	this.gc=[];
	this.rc = [];
	this.mc = [];
	this.ttt = []
	this.stage = [];
	this.m1 = [];
	this.m2 = [];
	this.m3 = [];
	this.m4 = [];
	this.mhc = [];
	this.sInt = [];
	this.id = 0;
}
Stage.prototype.getId = function() {
    return this.id;
}
Stage.prototype.setId = function(id){
    this.id = id;
}
Stage.prototype.getGC = function() {
    return this.gc;
}

Stage.prototype.setGC = function(array) {
    if (array.constructor === Array) {
        this.gc = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getRC = function() {
    return this.rc;
}
Stage.prototype.setRC = function(array) {
    if (array.constructor === Array) {
        this.rc = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getMC = function() {
    return this.mc;
}
Stage.prototype.setMC = function(array) {
    if (array.constructor === Array) {
        this.mc = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getTTT = function() {
    return this.ttt;
}
Stage.prototype.setTTT = function(array) {
    if (array.constructor === Array) {
        this.ttt = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getStage = function() {
    return this.stage;
}
Stage.prototype.setStage = function(array) {
    if (array.constructor === Array) {
        this.stage = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.setM1 = function(array) {
    if (array.constructor === Array) {
        this.m1 = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getM1 = function() {
    return this.m1;
}
Stage.prototype.setM2 = function(array) {
    if (array.constructor === Array) {
        this.m2 = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getM2 = function() {
    return this.m2;
}
Stage.prototype.setM3 = function(array) {
    if (array.constructor === Array) {
        this.m3 = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getM3 = function() {
    return this.m3;
}
Stage.prototype.setM4 = function(array) {
    if (array.constructor === Array) {
        this.m4 = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getM4 = function() {
    return this.m4;
}
Stage.prototype.setMHC = function(array) {
    if (array.constructor === Array) {
        this.mhc = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getMHC = function() {
    return this.mhc;
}
Stage.prototype.setSInt = function(array) {
    if (array.constructor === Array) {
        this.sInt = array;
    } else {
        alert("This is not an array");
    }
}
Stage.prototype.getSInt = function() {
    return this.sInt;
}

//Helpers

Stage.prototype.addSInt = function(sInt) {
    this.sInt.push(sInt);
}
Stage.prototype.addM1 = function(m1){
    this.m1.push(m1);
}
Stage.prototype.addM2 = function(m2) {
    this.m2.push(m2);
}
Stage.prototype.addM3 = function(m3) {
    this.m3.push(m3);
}
Stage.prototype.addM4 = function(m4) {
    this.m4.push(m4);
}
Stage.prototype.addMHC = function(mhc) {
    this.mhc.push(mhc);
}
Stage.prototype.toJSON = function() {
    var json = {};
    json.id = this.id;
    if (this.gc.length > 0) {
        json.gc = this.gc;
    }
    if (this.rc.length > 0) {
        json.rc = this.rc;
    }
    if (this.mc.length > 0) {
        json.mc = this.mc;
    }
    if (this.ttt.length > 0) {
        json.ttt = this.ttt;
    }
    if (this.stage.length>0) {
        json.stage = this.stage;
    }
    if (this.m1.length > 0) {
        json.m1 = this.m1;
    }
    if (this.m2.length > 0) {
        json.m2 = this.m2;
    }
    if (this.m3.length > 0) {
        json.m3 = this.m3;
    }
    if (this.m4.length > 0) {
        json.m4 = this.m4;
    }
    if (this.mhc.length > 0) {
        json.mhc = this.mhc;
    }
    if (this.sInt.length > 0) {
        json.sInt = this.sInt;
    }
    return json;
}